internal/controller/numaflowcontrollerrollout/numaflowcontrollerrollout_controller.go

```diff
@@ -811,124 +541,49 @@ func (r *NumaflowControllerRolloutReconciler) SetupWithManager(mgr ctrl.Manager)
+			&apiv1.NumaflowControllerRollout{}, handler.OnlyControllerOwner()), predicate.TypedResourceVersionChangedPredicate[*apiv1.NumaflowController]{})); err != nil {
+		return fmt.Errorf("failed to watch NumaflowController: %v", err)
 	}
 
 	return nil
 }
 
-// SplitYAMLToString splits a YAML file into strings. Returns list of yamls
-// found in the yaml. If an error occurs, returns objects that have been parsed so far too.
-func SplitYAMLToString(yamlData []byte) ([]string, error) {
-	d := yaml.NewYAMLOrJSONDecoder(bytes.NewReader(yamlData), 4096)
-	var objs []string
-	for {
-		ext := runtime.RawExtension{}
-		if err := d.Decode(&ext); err != nil {
-			if err == io.EOF {
-				break
-			}
-			return objs, fmt.Errorf("failed to unmarshal manifest: %v", err)
-		}
-		ext.Raw = bytes.TrimSpace(ext.Raw)
-		if len(ext.Raw) == 0 || bytes.Equal(ext.Raw, []byte("null")) {
-			continue
-		}
-		objs = append(objs, string(ext.Raw))
-	}
-	return objs, nil
+func (r *NumaflowControllerRolloutReconciler) updateNumaflowControllerRolloutStatus(ctx context.Context, nfcRollout *apiv1.NumaflowControllerRollout) error {
+	return r.client.Status().Update(ctx, nfcRollout)
 }
 
-func toUnstructuredAndApplyLabel(manifests []string, name string) ([]*unstructured.Unstructured, error) {
-	uns := make([]*unstructured.Unstructured, 0)
-	for _, m := range manifests {
-		obj := make(map[string]interface{})
-		err := yaml.Unmarshal([]byte(m), &obj)
-		if err != nil {
-			return nil, err
-		}
-		target := &unstructured.Unstructured{Object: obj}
-		err = kubernetes.SetLabel(target, common.LabelKeyNumaplaneInstance, name)
-		if err != nil {
-			return nil, err
-		}
-		uns = append(uns, target)
-	}
-	return uns, nil
+func (r *NumaflowControllerRolloutReconciler) updateNumaflowControllerRolloutStatusToFailed(ctx context.Context, nfcRollout *apiv1.NumaflowControllerRollout, err error) error {
+	nfcRollout.Status.MarkFailed(err.Error())
+	return r.updateNumaflowControllerRolloutStatus(ctx, nfcRollout)
 }
 
-func (r *NumaflowControllerRolloutReconciler) updateNumaflowControllerRolloutStatus(ctx context.Context, controllerRollout *apiv1.NumaflowControllerRollout) error {
-	return r.client.Status().Update(ctx, controllerRollout)
+func (r *NumaflowControllerRolloutReconciler) ErrorHandler(nfcRollout *apiv1.NumaflowControllerRollout, err error, reason, msg string) {
+	r.customMetrics.NumaflowControllerRolloutSyncErrors.WithLabelValues().Inc()
+	r.recorder.Eventf(nfcRollout, corev1.EventTypeWarning, reason, msg+" %v", err.Error())
 }
 
-func (r *NumaflowControllerRolloutReconciler) updateNumaflowControllerRolloutStatusToFailed(ctx context.Context, controllerRollout *apiv1.NumaflowControllerRollout, err error) error {
-	controllerRollout.Status.MarkFailed(err.Error())
-	return r.updateNumaflowControllerRolloutStatus(ctx, controllerRollout)
-}
+func generateNewNumaflowControllerDef(nfcRollout *apiv1.NumaflowControllerRollout) (*unstructured.Unstructured, error) {
```

Comment:
this could also return a `NumaflowController` type

Reply:
note: if we do decide to change these to `NumaflowController` type, I'm okay if it's a follow-up PR

Reply:
I see you needed to use Unstructured in order to use usde, so I think I'm good with it

Reply:
Same as above.